daemon/api_validate.go

```diff
@@ -0,0 +1,287 @@
+		if err != nil {
+			return InternalError(err.Error())
+		}
+		results[i] = validationSetResult{
+			AccountID: tr.AccountID,
+			Name:      tr.Name,
+			PinnedAt:  tr.PinnedAt,
+			Mode:      modeStr,
+			Sequence:  tr.Current,
+			Valid:     valid,
+		}
+	}
+
+	return SyncResponse(results, nil)
+}
+
+func getValidationSet(c *Command, r *http.Request, _ *auth.UserState) Response {
+	vars := muxVars(r)
+	accountID := vars["account"]
+	name := vars["name"]
+
+	if !asserts.IsValidAccountID(accountID) {
+		return BadRequest("invalid account ID %q", accountID)
+	}
+	if !asserts.IsValidValidationSetName(name) {
+		return BadRequest("invalid name %q", name)
+	}
+
+	query := r.URL.Query()
+
+	// sequence is optional
+	sequenceStr := query.Get("sequence")
+	var sequence int
+	if sequenceStr != "" {
+		var err error
+		sequence, err = strconv.Atoi(sequenceStr)
+		if err != nil {
+			return BadRequest("invalid sequence argument")
+		}
+		if sequence < 0 {
+			return BadRequest("invalid sequence argument: %d", sequence)
+		}
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	var tr assertstate.ValidationSetTracking
+	err := assertstate.GetValidationSet(st, accountID, name, &tr)
+	if err == state.ErrNoState || (err == nil && sequence != 0 && sequence != tr.PinnedAt) {
+		// TODO: not available locally, try to find in the store.
+		return validationSetNotFound(accountID, name, sequence)
+	}
+	if err != nil {
+		return InternalError("accessing validation sets failed: %v", err)
+	}
+
+	modeStr, err := modeString(tr.Mode)
+	if err != nil {
+		return InternalError(err.Error())
+	}
+	// TODO: evaluate against installed snaps
+	var valid bool
+	res := validationSetResult{
+		AccountID: tr.AccountID,
+		Name:      tr.Name,
+		PinnedAt:  tr.PinnedAt,
+		Mode:      modeStr,
+		Sequence:  tr.Current,
+		Valid:     valid,
+	}
+	return SyncResponse(res, nil)
+}
+
+type validationSetApplyRequest struct {
+	Action   string `json:"action"`
+	Mode     string `json:"mode"`
+	Sequence int    `json:"sequence,omitempty"`
+}
+
+func applyValidationSet(c *Command, r *http.Request, _ *auth.UserState) Response {
+	vars := muxVars(r)
+	accountID := vars["account"]
+	name := vars["name"]
+
+	if !asserts.IsValidAccountID(accountID) {
+		return BadRequest("invalid account ID %q", accountID)
+	}
+	if !asserts.IsValidValidationSetName(name) {
+		return BadRequest("invalid name %q", name)
+	}
+
+	var req validationSetApplyRequest
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&req); err != nil {
+		return BadRequest("cannot decode request body into validation set action: %v", err)
+	}
+	if decoder.More() {
+		return BadRequest("extra content found in request body")
+	}
+	if req.Sequence < 0 {
+		return BadRequest("invalid sequence argument: %d", req.Sequence)
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if req.Action == "forget" {
+		return forgetValidationSet(st, accountID, name, req.Sequence)
```

Comment:
(nitpick) I personally would probably split the rest of the function into a helper similar to forgetValidationSet and make the if req.Action() a switch/case. But this here is also fine.

Reply:
Done